File: 300/355 - Design Twitter.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.TreeSet;

public class Twitter {
    
    int count;
    HashMap<Integer, HashSet<Integer>> follows;
    HashMap<Integer, TreeSet<Tweet>> tweets;
    TreeSet<Tweet> topTen;
    TweetComparator<Tweet> tweetComparator;
    

    /** Initialize your data structure here. */
    public Twitter() {
        this.count = 0;
        this.follows = new HashMap<Integer, HashSet<Integer>>();
        this.tweets= new HashMap<Integer, TreeSet<Tweet>>();
        this.tweetComparator = new TweetComparator<Tweet>(false);
        this.topTen = new TreeSet<Tweet>(new TweetComparator<Tweet>(true));
    }
    
    /** Compose a new tweet. */
    public void postTweet(int userId, int tweetId) {
        follow(userId, userId);
        
        if (! this.tweets.containsKey(userId))
            this.tweets.put(userId, new TreeSet<Tweet>(this.tweetComparator));
        
        this.tweets.get(userId).add(new Tweet(this.count++, tweetId));
    }
    
    /** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
    public List<Integer> getNewsFeed(int userId) {
        List<Integer> ret = new ArrayList<Integer>();
        
        if (this.follows.containsKey(userId)) {
            
            for (Integer followee : this.follows.get(userId)) {
                
                if (this.tweets.containsKey(followee)) {
                    
                    for (Tweet tweet : this.tweets.get(followee)) {
                        
                        if (this.topTen.size() < 10)
                            this.topTen.add(tweet);
                        else {
                            
                            if (this.topTen.first().getIndex() > tweet.getIndex())
                                break;
                            else {
                                this.topTen.pollFirst();
                                this.topTen.add(tweet);
                            }
                        }
                    }
                }
            }
        }
        
        for (Tweet tweet : this.topTen) {
            ret.add(0, tweet.getId());
        }
        
        this.topTen.clear();
        
        return ret;
    }
    
    /** Follower follows a followee. If the operation is invalid, it should be a no-op. */
    public void follow(int followerId, int followeeId) {
        if (! this.follows.containsKey(followerId))
            this.follows.put(followerId, new HashSet<Integer>());
        
        this.follows.get(followerId).add(followeeId);
    }
    
    /** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
    public void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId || ! this.follows.containsKey(followerId))
            return;
        
        this.follows.get(followerId).remove(followeeId);
    }
    
    private class Tweet {
        
        private int index;
        private int id;
        
        public Tweet(int index, int id) {
            this.index = index;
            this.id = id;
        }
        
        public int getIndex() {
            return this.index;
        }
        
        public int getId() {
            return this.id;
        }
    }
    
    private class TweetComparator<Tweet> implements Comparator<Tweet> {
        
        private boolean isAscending;
        
        public TweetComparator(boolean isAscending) {
            this.isAscending = isAscending;
        }
        
        @Override
        public int compare(Twitter.Tweet t1, Twitter.Tweet t2) {
            return this.isAscending ? t1.getIndex() - t2.getIndex() : t2.getIndex() - t1.getIndex();
        }
    }
}

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */

File: csharp/355.design-twitter.cs
/*
 * @lc app=leetcode id=355 lang=csharp
 *
 * [355] Design Twitter
 */

using System.Collections.Generic;
using System.Linq;

public class Twitter {

    private IDictionary<int, ISet<int>> Followings;
    private IDictionary<int, LinkedList<(int, int)>> Tweets;
    private int Time;

    /** Initialize your data structure here. */
    public Twitter() {
        Followings = new Dictionary<int, ISet<int>>();
        Tweets = new Dictionary<int, LinkedList<(int, int)>>();
        Time = int.MinValue;
    }
    
    /** Compose a new tweet. */
    public void PostTweet(int userId, int tweetId) {
        LinkedList<(int, int)> list;
        if (!Tweets.ContainsKey(userId)) {
            list = new LinkedList<(int, int)>();
            Tweets[userId] = list;
        } else {
            list = Tweets[userId];
        }
        list.AddFirst((Time++, tweetId));
    }
    
    /** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
    public IList<int> GetNewsFeed(int userId) {
        if (!Followings.ContainsKey(userId)) {
            Followings[userId] = new HashSet<int>();
        }
        SortedList<int, int> ret = new SortedList<int, int>(10, Comparer<int>.Create((a, b) => b.CompareTo(a)));
        foreach (LinkedList<(int, int)> list in Followings[userId].Union(Enumerable.Repeat(userId, 1)).Where(Tweets.ContainsKey).Select(u => Tweets[u])) {
            foreach ((int, int) entry in list) {
                int time = entry.Item1, tweet = entry.Item2;
                if (ret.Count < 10) {
                    ret.Add(time, tweet);
                    continue;
                }
                if (time < ret.Last().Key) {
                    break;
                }
                ret.RemoveAt(9);
                ret.Add(time, tweet);
            }
        }
        return ret.Values;
    }
    
    /** Follower follows a followee. If the operation is invalid, it should be a no-op. */
    public void Follow(int followerId, int followeeId) {
        ISet<int> set;
        if (!Followings.ContainsKey(followerId)) {
            set = new HashSet<int>();
            Followings[followerId] = set;
        } else {
            set = Followings[followerId];
        }
        set.Add(followeeId);
    }
    
    /** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
    public void Unfollow(int followerId, int followeeId) {
        if (Followings.ContainsKey(followerId)) {
            Followings[followerId].Remove(followeeId);
        }
    }
}

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.PostTweet(userId,tweetId);
 * IList<int> param_2 = obj.GetNewsFeed(userId);
 * obj.Follow(followerId,followeeId);
 * obj.Unfollow(followerId,followeeId);
 */

